src/polly/src/event_manager.rs

```diff
@@ -61,6 +61,68 @@ pub trait Subscriber {
 
     /// Returns a list of `EpollEvent` that this subscriber is interested in.
     fn interest_list(&self) -> Vec<EpollEvent>;
+
+    /// Function used by the default implementation of `pause()` to decide if it should
+    /// unregister `Subscriber` events from `EventManager`.
+    ///
+    /// Override this function with custom logic to control the default implementation
+    /// of `pause()`.
+    fn unregister_events_on_pause(&self) -> bool {
```

Comment:
Adding the pause/resume specific functions in polly looks incompatible with the initial polly design which was intended as a reusable event management compoennt. Changing the interface of polly at this level would also make it really hard to switch to event_manager (#2154) in the future as it unlikely that such a functionality will be implemented at this level.

Would it be possible to abstract this at a different level? If boiler plate code is the concern that generated these changes to happen at the interface level, it might make sense to explore other technical solutions as well. WDYT?

Reply:
I agree with your concerns, but it all boils down with whether we want `polly` or the upstream `EventManager` to be able to offer control to their underlying `Subscribers` over registering and un-registering their (or other's) events in the current `EventManager` loop/`run()`.

### *The Scenario*

Let me try to set a clear scenario and we can definitely explore other technical solutions for handling it:
- there are multiple registered `Subscribers`,
- one of them is the main one, the driver if you will, or in our case the `Vmm`,
- at some point the driver/vmm, wants one of the other `subscribers`, in our case _a device_, to register or unregister their events **based on the device's internal state and/or data**

### Options (identified so far)

To do this I can identify two options:
1. Using the existing `polly/event_manager` interface:
   1. The `vmm` downcasts whatever handle it has for the device (`trait BusDevice` or some `trait Bla` that has the ability to navigate down to the concrete type; `trait Subscriber` for example doesn't currently have that ability) all the way down to the concrete type.
   2. Having the concrete type, it inspects the device's internal state/data, and based on that it decides whether to register/unregister that device to/from `event_manager`.
   3. To be able to call `event_manager.register()` we need to upcast the concrete type to `Arc<Mutex<dyn Subscriber>>` (needed by `register()`).
   4. To do this upcast we need to add *a lot of god ugly boiler-plate* conversions to get from `MutexGuard<ConcreteType>` to `Arc<Mutex<dyn Subscriber>>`.
2. Enhancing `trait Subscriber` interface to allow native/internal handling of the scenario:
   1. `Subscriber` offers an interface for the device to do it itself. We can add better, generic names, but for this example let's stick with `pause/resume`.
   2. Unlike `1. i`, when working through the trait, there is no need to do any downcasting, so `vmm` simply calls `device.pause()`. The device implementation natively runs on the `Self` concrete type.
   3. Unlike `1. iii`, we're already working with `Arc<Mutex<dyn Subscriber>>` to start with, so we can easily use `event_manager.register()` in `device.resume()` with no dynamic up or down casting necessary.
3. (Hacky) Existing `polly` interface: use _fake events_ in every device and manually call `process()`.
   - Works same as **option 2** but instead of adding `pause/resume` we define fake `EpollEvents` that decode as `pause` or `resume` commands.
   - `vmm` calls `device.process()` and pushes one of these fake events.
   - when defining/choosing fake event values, we need to take care not to have conflicts between real events (`FDs`) and our fake ones.

### Personal preference

Obviously *option 2* since that's the one that made it to the PR 😄 
You can also see *option 1* as work-in-progress in [this branch](https://github.com/acatangiu/firecracker/tree/pause-emulation), but it is incomplete as I just gave up on all the up and down casting somewhere along the way 😆 and looked for alternatives.

Aside from that, option 1 also has the downside of leaking or intentionally exposing internal logic as needed by step `1. ii`. The device is not independent, the `vmm` is the one to externally call `register/unregister` for the device; the device itself can't do that outside `trait Subscriber` functions without taking `&mut EventManager` outside of its `impl Subscriber`.

#### Option 2 interface improvements?

1. More generic naming, maybe not `pause/resume`?
2. Don't offer default impl for `pause/resume`, then you don't need other four control functions that control the default impl. Downside is more boilerplate code.
3. Probably many more 🙈 

### The big question

Like I said in the beginning, when evaluating options, we need to first answer this question:

Is *The Scenario* something that more users of `EventManager` are interested in? Or is it something `Firecracker` specific?

My money is on the former.